proposals/kapp-controller/004-dependency-resolution/README.md

```diff
@@ -0,0 +1,621 @@
+
+    - name: dep4
+      wantedPackageRef:
+        refName: pkg4.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef: 
+        name: dep-pkgi.this-is-another-random-string
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: microservice-v1
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: other-pkg.pkg5.test.carvel.dev
+        namespace: ns1
+      conditions:
+      - type: FoundPackageInstall
+
+    - name: microservice-v2
+      wantedPackageRef:
+        refName: pkg5.test.carvel.dev
+        versionSelection:
+          constraints: 0.2.0
+      packageInstallRef:
+        name: pkg-demo.microservice-v2
+        namespace: default
+      conditions:
+      - type: OwnedPackageInstall
+
+    - name: optional-dependency
+      wantedPackageRef:
+        refName: pkg6.test.carvel.dev
+        versionSelection:
+          constraints: 0.6.0
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-wrong-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: "1.0.1" # one on the cluster is 0.1.1
+      packageInstallRef:
+        name: pkg-demo.pkg1.test.carvel.dev
+      conditions:
+      - type: FoundPackageInstall
+      - type: FoundPackageInstallRequiresDifferentVersion
+
+    - name: dep-with-unknown-package
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+      conditions:
+      - type: UnknownPackage
+
+    - name: dep-with-package-with-high-version
+      wantedPackageRef:
+        refName: pkg1.test.carvel.dev
+        versionSelection:
+          constraints: ">4.0.0" # cant calculate exact version in this system
+      conditions:
+      - type: UnknownPackageVersion
+
+```
+
+Example of a PackageInstall that is automatically created
+```yaml
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
+spec:
+    serviceAccountName: default-ns-sa
+  packageRef:
+    refName: pkg3.test.carvel.dev
+    versionSelection:
+      constraints: 1.0.0
+  values:
+    secretRef:
+      name: pkg3.dev-values
+...
+```
+
+#### Dependency resolution
+The dependency resolution will be done one PackageInstall at a time, basically each PackageInstall will only create the PackageInstalls that it directly depends on.
+
+ex:
+```
+When pkg depends on pkg1 and pkg3, while pkg1 depends on pkg2
+kapp-controller when it is processing pkg it will only
+create PackageInstalls for pkg1 and pkg3.
+
+pkg
+  -> pkg1
+      -> pkg2
+  -> pkg3
+```
+
+For future versions we can contemplate the possibility of each PackageInstall checking if the dependencies it has can be fulfilled with the Packages present in the Cluster.
```

Comment:
This means the proposal does not cover runtime dependencies. If pkg 1 --> pkg 2 (v1.0.0), and the cluster already contains pkg2 (v1.1.0) installed - do we just fail (even though it is just a minor bump and is expected to be backwards compatible)?

Reply:
This sentence should have been removed (my bad). The following condition in the status already covers this scenario where if an already installed package satisfies the dependency then we don't install the package again
```yaml   
- name: microservice-v1
      wantedPackageRef:
        refName: pkg5.test.carvel.dev
        versionSelection:
          constraints: 0.2.0
      packageInstallRef:
        name: other-pkg.pkg5.test.carvel.dev
        namespace: ns1
      conditions:
      - type: FoundPackageInstall
 ```

Reply:
> If pkg 1 --> pkg 2 (v1.0.0), and the cluster already contains pkg2 (v1.1.0) installed - do we just fail (even though it is just a minor bump and is expected to be backwards compatible)

We want to add a `installationType` field on the package for this (check package crd). This will determine if the package can be installed again or not. If it cannot be installed again and the version installed on the cluster doesn't satisfy the version constraints provided on the package for the dependency, then we will fail with 
```yaml
    - name: dep-with-wrong-version
      wantedPackageRef:
        refName: pkg1.test.carvel.dev
        versionSelection:
          constraints: "1.0.1" # one on the cluster is 0.1.1
      packageInstallRef:
        name: pkg-demo.pkg1.test.carvel.dev
      conditions:
      - type: FoundPackageInstall
      - type: FoundPackageInstallRequiresDifferentVersion
  ```

Reply:
> If pkg 1 --> pkg 2 (v1.0.0), and the cluster already contains pkg2 (v1.1.0) installed - do we just fail (even though it is just a minor bump and is expected to be backwards compatible)
@praveenrewar  since we support range in the versionSelection.constraints, then I think users can get this behaviour if they choose to.